presto-main/src/test/java/com/facebook/presto/operator/TestAnnotationEngine.java

```diff
@@ -0,0 +1,884 @@
+import com.facebook.presto.type.TypeRegistry;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.testng.annotations.Test;
+
+import java.util.List;
+
+import static com.facebook.presto.metadata.Signature.typeVariable;
+import static com.facebook.presto.operator.AnnotationBasedOperators.ComplexParametricScalarFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.ConstructorInjectionScalarFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.ExactAggregationFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.ExplicitSpecializedAggregationFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.MultiScalarFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.ParametricScalarFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.SimpleInjectionScalarFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.StaticMethodScalarFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.WithNullableComplexArgScalarFunction;
+import static com.facebook.presto.operator.AnnotationBasedOperators.WithNullablePrimitiveArgScalarFunction;
+import static com.facebook.presto.operator.aggregation.AggregationFromAnnotationsParser.parseFunctionDefinition;
+import static com.facebook.presto.operator.aggregation.AggregationFromAnnotationsParser.parseFunctionDefinitions;
+import static com.facebook.presto.spi.type.StandardTypes.ARRAY;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class TestAnnotationEngine
+{
+    @Test
+    public void testSimpleExactAggregationParse()
+            throws Exception
+    {
+        Signature expectedSignature = new Signature(
+                "simple_exact_aggregate",
+                FunctionKind.AGGREGATE,
+                DoubleType.DOUBLE.getTypeSignature(),
+                ImmutableList.of(DoubleType.DOUBLE.getTypeSignature()));
+
+        ParametricAggregation aggregation = parseFunctionDefinition(ExactAggregationFunction.class);
+        assertEquals(aggregation.getDescription(), "Simple exact aggregate description");
+        assertTrue(aggregation.isDeterministic());
+        assertEquals(aggregation.getSignature(), expectedSignature);
+        ParametricImplementationsGroup<AggregationImplementation> implementations = aggregation.getImplementations();
+        assertEquals(implementations.getExactImplementations().size(), 1);
+        assertEquals(implementations.getGenericImplementations().size(), 0);
+        assertEquals(implementations.getSpecializedImplementations().size(), 0);
```

Comment:
Add a helper method that does

```
assertImplementationCount(implementations, int exact, int specialized, int generic)
{
    assertEquals(implementations.getExactImplementations().size(), exact);
    assertEquals(implementations.getSpecializedImplementations().size(), specialized);
    assertEquals(implementations.getGenericImplementations().size(), generic);
}
```

If you see other opportunities to reduce boilerplate, consider taking it.